### PR TITLE
Initial commit for code changes to move between past and next levels

### PR DESCRIPTION
**Pull request by otaiwo70**
feature will allow player move through level 

levels can be past and next levels

Next level represented by N and past level represented by R
Below are 2 levels

1st level
XXX
X  N
X  X
XXX

2nd Level
XXX
R  X
X  N
XXX

### CODE REVIEW

File: src/main/java/parser/LevelMove.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		gameEngine.getLevelCreator().createLevel(gameEngine, this.levelNum);
+		// this.levelCreator = gameEngine.getLevelCreator();
+		// this.levelCreator.createLevel(gameEngine, this.levelNum);
+	}
+
+	public void nextLevel(GameEngine gameEngine) {
+		if (levelNum < levelMax)
+			this.levelNum = this.levelNum + 1;
+		gameEngine.getLevelCreator().createLevel(gameEngine, this.levelNum);
+		// this.levelCreator.createLevel(gameEngine, this.levelNum);
```
Comment (borisv13) [inline src/main/java/parser/LevelMove.java]: delete comment

Comment (otaiwo70) [reply src/main/java/parser/LevelMove.java]: confirm comment removed

File: src/main/java/parser/LevelMove.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	public void pastLevel(GameEngine gameEngine) {
+		if (levelNum > levelMin)
+			this.levelNum = this.levelNum - 1;
+		gameEngine.getLevelCreator().createLevel(gameEngine, this.levelNum);
+		// this.levelCreator = gameEngine.getLevelCreator();
+		// this.levelCreator.createLevel(gameEngine, this.levelNum);
+	}
+
+	public void nextLevel(GameEngine gameEngine) {
+		if (levelNum < levelMax)
```
[the remark this answers is not in the document]
Comment (otaiwo70) [reply src/main/java/parser/LevelMove.java]: confirm curly braces removed

File: src/test/java/parser/LevelMoveTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	@Test
+	public void set_and_get_level_max() {
+		levelMove.setLevelMax(3);
+		int actual = levelMove.getLevelMax();
+		assertEquals(3, actual);
+	}
+
+//	@Test
```
Comment (borisv13) [inline src/test/java/parser/LevelMoveTest.java]: Delete comments

Comment (otaiwo70) [reply src/test/java/parser/LevelMoveTest.java]: Sorry missed it comment removed

**Comment by otaiwo70** [comment]
Added test i missed to cover sonar covergae

**Comment by otaiwo70** [comment]
Acceptance Test added for Cucumber

File: src/main/java/engine/GameEngine.java [hunk 4]
```diff
@@ -94,6 +97,12 @@ private void movePlayer(int xDiff, int yDiff) {
 				getPlayerYCoordinate() + yDiff);
 		if (attempedLocation.equals(TileType.PASSABLE)) {
 			setPlayer(getPlayerXCoordinate() + xDiff, getPlayerYCoordinate() + yDiff);
+		} else if (attempedLocation.equals(TileType.PAST_LEVEL)) {
+			setPlayer(getPlayerXCoordinate() + xDiff, getPlayerYCoordinate() + yDiff);
+			this.levelMove.pastLevel(this);
+		}else if (attempedLocation.equals(TileType.NEXT_LEVEL)) {
```
Comment (borisv13) [inline src/main/java/engine/GameEngine.java]: **Turn your code formatter on!!! From Assignment 1**
There should be a space after the { before the else.
Please go through every file you changed and make sure to format each and every one after you turn the formatter back on.

File: src/main/java/engine/GameEngine.java [hunk 5]
```diff
@@ -114,6 +114,10 @@ public boolean isExit() {
 		return exit;
 	}
 
+	public LevelMove getLevelMove() {
```
Comment (borisv13) [inline src/main/java/engine/GameEngine.java]: I don't see this being used in prod anywhere. Am I missing it? Or does this method just exist for a test case? Cause that should never happen.

File: src/test/java/engine/GameEngineTest.java [hunk 6]
```diff
@@ -79,4 +82,61 @@ public void set_and_get_exit() {
...
 
+	@Test
+	public void get_current_level_move() {
+		levelMove = gameEngine.getLevelMove();
+		int actual = levelMove.getLevelNum();
+		assertThat(actual, equalTo(ONE));
+	}
+
+	@Test
+	public void check_player_can_move_to_past_level_tile() {
```
Comment (borisv13) [inline src/test/java/engine/GameEngineTest.java]: These 4 tests demonstrate that you did not get the lesson from the in-class experience when we implemented movement.
A Unit test where you have to mock out every single Tile is obnoxious. The setup is huge. The test is unreadable. You should not write it - it's a bad test. You should instead create a real level, move the player, and assert what happened.
This lesson was the whole point of that in-class exercise.

Delete these 4 tests and test this behavior in the BETTER, FASTER approach.

File: src/test/java/parser/LevelMoveNextStepDefs.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+	private GameEngine gameEngine;
+	private LevelMove levelMove;
+	private LevelCreator levelCreator;
+	private static final int ONE = 1;
+	private static final int TWO = 2;
+	private static final int THREE = 3;
+
+	@Given("^grid containing next level (\\d+) is:$")
+	public void grid_containing_next_level(int arg1, DataTable arg2) throws Throwable {
```
Comment (borisv13) [inline src/test/java/parser/LevelMoveNextStepDefs.java]: arg1 and arg2 are great variable names :)

File: src/test/resources/features/parser/LevelMoveNext.feature [hunk 8]
```diff
@@ -0,0 +1,15 @@
+@IntegrationTest
+Feature: Move the player to next level
+
+  Background: 
+    Given grid containing next level 2 is:
```
Comment (borisv13) [inline src/test/resources/features/parser/LevelMoveNext.feature]: I don't understand the English here. "grid containing next level 2"
1. This is unclear
2. Are you saying the next level is 2 and that is the next level?
3. Then why are you asserting that the next level is 3 in the Then?

File: src/test/resources/features/parser/LevelMovePast.feature [hunk 9]
```diff
@@ -0,0 +1,15 @@
+@IntegrationTest
+Feature: Move the player to next level
+
+  Background: 
+    Given grid containing past level 3 is:
```
Comment (borisv13) [inline src/test/resources/features/parser/LevelMovePast.feature]: Same as previous.

File: src/main/java/engine/GameEngine.java (hunk 4 above)
Comment (otaiwo70) [reply src/main/java/engine/GameEngine.java]: Code has been re-formatted, formatter was on but configuring options needed to added

File: src/test/java/engine/GameEngineTest.java (hunk 6 above)
Comment (otaiwo70) [reply src/test/java/engine/GameEngineTest.java]: Extracted these out into a separate unit test with actual game utilized, the mock was utilized initially as i felt i needed to go with the flow of how GameEngineTest was structured initially

File: src/test/java/parser/LevelMoveNextStepDefs.java (hunk 7 above)
Comment (otaiwo70) [reply src/test/java/parser/LevelMoveNextStepDefs.java]: args have been updated to provide clarity

File: src/test/resources/features/parser/LevelMoveNext.feature (hunk 8 above)
Comment (otaiwo70) [reply src/test/resources/features/parser/LevelMoveNext.feature]: Sentence updated to provide clalrity

File: src/test/resources/features/parser/LevelMovePast.feature (hunk 9 above)
Comment (otaiwo70) [reply src/test/resources/features/parser/LevelMovePast.feature]: Sentence updated to provide clalrity

File: src/main/java/engine/GameEngine.java (hunk 5 above)
Comment (otaiwo70) [reply src/main/java/engine/GameEngine.java]: Was supposed to use in movePlayer but forgot, updated the code as intended

File: src/main/java/engine/GameEngine.java [hunk 10]
```diff
@@ -6,102 +6,119 @@
...
+				this.getPlayerYCoordinate() + yDiff);
 		if (attempedLocation.equals(TileType.PASSABLE)) {
-			setPlayer(getPlayerXCoordinate() + xDiff, getPlayerYCoordinate() + yDiff);
+			this.setPlayer(this.getPlayerXCoordinate() + xDiff, this.getPlayerYCoordinate() + yDiff);
+		} else if (attempedLocation.equals(TileType.PAST_LEVEL)) {
+			this.setPlayer(this.getPlayerXCoordinate() + xDiff, this.getPlayerYCoordinate() + yDiff);
+			this.getLevelMove().pastLevel(this);
+		} else if (attempedLocation.equals(TileType.NEXT_LEVEL)) {
+			this.setPlayer(this.getPlayerXCoordinate() + xDiff, this.getPlayerYCoordinate() + yDiff);
+			this.getLevelMove().nextLevel(this);
```
Comment (borisv13) [inline src/main/java/engine/GameEngine.java]: You don't need a getter for an instance of an object local to itself:
this.levelMove.nextLevel(this);
Now you don't have to expose LevelMove to the rest of the world.
Remember: data encapsulation.

Comment (otaiwo70) [reply src/main/java/engine/GameEngine.java]: The method has been removed and src and test code using the method refactored to utilize method in LevelMove

**Pull request closed without merging**